docs/design/scheduling-reason.md

```diff
@@ -0,0 +1,46 @@
+# Scheduling Reason
+
+[@eggiter](https://github.com/eggiter); Aug 13, 2021
+
+## Table of Contents
+
+* [Table of Contents](#table-of-contents)
+* [Motivation](#motivation)
+* [Solution](#solution)
+
+## Motivation
+
+- Currently, the reason and message in `pod.status.conditions["PodScheduled"]` and `podgroup.status.conditions["Unschedulable"]` is **NOT** informative, for example `x/x tasks in gang unschedulable: pod group is not ready, x Pending, x minAvailable`.
+- Users want to know which one breaks the scheduling cycle and why is that.
+
+## Solution
+
+- Introducing two extra scheduling reasons in PodScheduled PodCondition:
+  + `Schedulable`: means that the scheduler can schedule the pod right now, but other pods in podgroup can not be scheduled
+  + `Undetermined`: means that the scheduler dose not schedule the pod yet, for example due to unschedulable pod already occurred.
+
+- Case:
+  + 3 nodes: node1, node2, node3;
+  + 6 tasks in the PodGroup(`minAvailable = 6`);
+    + only 1 task(Task6) is unschedulable and other 5 tasks(Task1 ~ 5) can be scheduled, thus the whole job is unscheduable.
+  
+- Current information:
+  +  |Tasks|Reason|Message|
+     |-|-|-|
+     |PodGroup |NotEnoughResources|6/6 tasks in gang unschedulable: pod group is not ready, 6 Pending, 6 minAvailable |
+     |Task1 ~ 5|Unschedulable|6/6 tasks in gang unschedulable: pod group is not ready, 6 Pending, 6 minAvailable|
+     |Task6|Unschedulable| all nodes are unavailable: 1 plugin InterPodAffinity predicates failed node(s) didn't match pod affinity/anti-affinity, node(s) didn't match pod affinity rules, 2 node(s) resource fit failed.|
+
+- Improved information:
+  + |Tasks|Reason|Message|
+    |-|-|-|
+    |PodGroup |(same)|6/6 tasks in gang unschedulable: pod group is not ready, 6 Pending, 6 minAvailable; **Pending: 1 Unschedulable, 2 Undetermined, 3 Schedulable** |
```

Comment:
BTW, l think 6/6 is not correctly calculated, can you please help check it? l think the definition of this is `minAvailable - canbescheduled pods/total pods" ?

Reply:
Should we take allocated but rollbacked tasks as `ReadyTask` used in plugin `gang`? 
+ If not, we'd better to remain this way. 
+ If so, the message in podgroup condition should be: 
    - "3/6 tasks in gang unschedulable: pod group is not ready, 6 Pending, 6 minAvailable; **Pending: 1 Unschedulable, 2 Undetermined, 3 Schedulable**"
    -  or "3/6 tasks in gang unschedulable: pod group is not ready, 3 Pending, 3 Allocated, 6 minAvailable; **Pending: 1 Unschedulable, 2 Undetermined**"?

Reply:
My understanding from your implementation, pending is the state of the current task, and Unschedulable/Undetermined/Schedulable corresponds to the result of the intermediate state before commit/discard, so I mean is it possible to change the error message to be more friendly?

Reply:
My opinion e.g: "3/6 tasks in gang unschedulable/undetermined: pod group is not ready, 6 total, 6 minAvailable; 1 Unschedulable, 2 Undetermined, 3 Schedulable"

Reply:
sounds reasonable